File: app/models/catalog.py
import uuid

from pgvector.sqlalchemy import Vector
from sqlalchemy import ForeignKey, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.llm.embeddings import EMBED_DIM


class CatalogItem(Base):
    __tablename__ = "catalog_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    business_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("businesses.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(nullable=False)
    unit: Mapped[str] = mapped_column(nullable=False)

    tiers: Mapped[list["CatalogItemTier"]] = relationship(
        order_by="CatalogItemTier.min_qty",
        cascade="all, delete-orphan",
    )

    embedding: Mapped[list[float]] = mapped_column(Vector(EMBED_DIM), nullable=False)


class CatalogItemTier(Base):
    __tablename__ = "catalog_item_tiers"
    __table_args__ = (UniqueConstraint("catalog_item_id", "min_qty"),)

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    catalog_item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("catalog_items.id", ondelete="CASCADE"), nullable=False
    )
    min_qty: Mapped[int] = mapped_column(nullable=False)
    unit_price: Mapped[int] = mapped_column(nullable=False)

File: frontend/src/lib/catalog.ts
import { api } from "./api";

export type CatalogItemTier = {
  id: string;
  min_qty: number;
  unit_price: number;
};

export type CatalogItem = {
  id: string;
  name: string;
  unit: string;
  tiers: CatalogItemTier[];
};

export type CatalogItemInput = {
  name: string;
  unit: string;
  tiers: { min_qty: number; unit_price: number }[];
};

export const catalogApi = {
  list: () => api.get<CatalogItem[]>("/api/v1/catalog/items"),
  create: (item: CatalogItemInput) => api.post<CatalogItem>("/api/v1/catalog/items", item),
  update: (id: string, item: CatalogItemInput) =>
    api.put<CatalogItem>(`/api/v1/catalog/items/${id}`, item),
  remove: (id: string) => api.delete<void>(`/api/v1/catalog/items/${id}`),
};
